orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -2494,6 +2494,24 @@ Example 4:
            {{"X_1"}, "Cos", {"X"}},
            {{"dX"}, "Mul", {"X_1", "dY"}}}));
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(CosGrad)
+    .SetDomain(kOnnxDomain)
+    .SinceVersion(9)
+    .SetSupportLevel(OpSchema::SupportType::EXPERIMENTAL)
+    .SetDoc("Gradient function for Cos")
+    .AllowUncheckedAttributes()
+    .Input(0, "dY", "Cos output's grad", "T")
+    .Input(1, "X", "Input tensor", "T")
+    .Output(0, "dX", "Cos input's grad", "T")
+    .TypeConstraint(
+        "T",
+        {"tensor(float16)", "tensor(float)", "tensor(double)"},
+        "Constrain input and output types to all numeric tensors.")
+    .FunctionBody(ONNX_NAMESPACE::FunctionBodyHelper::BuildNodes(
+        {// nodes: {outputs, op, inputs, attributes}
+          {{"X_1"}, "Sin", {"X"}},
+          {{"dX"}, "Mul", {"X_1", "dY"}}}));
+
```

Comment:
best to add type and shape inference function. I believe if one is not present we run through the function body to infer the type and shape of the output

Reply:
For gradient Ops normally we can skip adding the shape and type inference, because our gradient builder will use the corresponding input's shape and type to set the output shape and type to the graph by default. Since such Ops is used during gradient builder only, adding shape and type inference is not necessary, and need effort to maintain (if there is bug...).

Reply:
@prathikr  : Can we capture this as a comment in the training_op_defs.cc... may be at the very top.